design/one-pager-consuming-k8s-clusters.md

```diff
@@ -0,0 +1,291 @@
+is appropriate for Kubernetes clusters to be modeled in the same manner as other
+managed services due to the fact that they are presented by the provider in the
+same fashion. However, in Crossplane, Kubernetes clusters are unique to other
+managed service offerings in that they can also be used as a target for running
+workloads.
+
+Workloads are provisioned using the `KubernetesApplication` type. It is
+namespace-scoped and is scheduled to a target cluster using labels to match a
+`KubernetesCluster` claim in its namespace. Because managed resources may only
+be bound to a single claim, the claim is namespaced, and `KubernetesApplication`
+resources can only be scheduled to `KubernetesCluster` claims in their
+namespace, any Kubernetes managed resource offering may only be consumed (i.e.
+deployed to) from a single namespace.
+
+In order to fully conceptualize the usage of Kubernetes clusters in Crossplane,
+it is helpful to separate the notions of a **Kubernetes cluster as a managed
+service** and a **Kubernetes cluster as a target for deploying workloads**. With
+this distinction made, we can assert that this proposal in no way modifies the
+managed service notion of a Kubernetes cluster, but instead formalizes this
+concept of a Kubernetes cluster as a deployment target.
+
+## Goals
+
+Because it is unlikely that Crossplane users will always want a 1-to-1
+namespace-to-cluster consumption model, it is desirable to support deploying
+workloads in any namespace to any Kubernetes cluster. However, it is also
+desirable to continue to treat managed Kubernetes services as a managed resource
+that is supplied by cloud providers. For this reason, any solution should be
+additive to the existing model.
+
+As a secondary goal for this design, it should be possible to use existing
+Kubernetes clusters (wherever they may be running) that were not provisioned by
+Crossplane as targets for deploying workloads. This concept has frequently been
+referred to as a "Bring Your Own Cluster" scenario.
+
+## Proposal
+
+This proposal can be separated into two main categories:
+1. The implementation of a new `KubernetesTarget` custom resource and
+   corresponding controller.
+1. The modification of the scheduling behavior for `KubernetesApplication`.
+
+### KubernetesTarget
+
+The purpose of the `KubernetesTarget` resource is to effectively "publish"
+Kubernetes clusters for usage in a namespace. It is namespace-scoped and can
+only reference Kubernetes cluster managed resources or the Kubernetes `Provider`
+type, both of which are cluster-scoped. The creation of a `KubernetesTarget`
+resource in a namespace will be watched by a single new secret propagation
+controller, which will get the `Secret` referenced by either the Kubernetes
+cluster managed resource or Kubernetes `Provider` and propagate it to the
+namespace of the `KubernetesTarget`. A local object reference to the `Secret`
```

Comment:
I see your point here. However, couldn't a similar situation occur even if the cluster was bound? Is it just that the act of claiming the resource feels stronger than just having a `KubernetesTarget` resource that point to it?

Reply:
My understanding is that the thinking (at the time I commented) was that `KubernetesTarget` would cause a secret to be propagated from the (e.g.) `GKECluster` to its namespace. This secret would need to be propagated for a `KubernetesTarget` to actually be used to deploy a `KubernetesApplication`. So if the secret was only propagated if and when the `GKECluster` was bound, then no I don't think a similar situation could occur.

Reply:
So in the scenario above, let’s say that App team B claims the managed resource before any KubernetsTarget is created. It isn’t now bound. However, this provides no assurance to App team B that another team (i.e. App team A) will not also start to utilize the cluster via a KubernetesConnection, which can now be used with the cluster because it is bound. It does provide assurance that when App team B first claimed this resource that no other team was currently using it. Is this what you are communicating here? I may still not be fully grasping but your intent.

Reply:
The technical implementation section now includes the requirement that a managed resource referenced by a `KubernetesTarget` must be `Bound` for the connection `Secret` to be propagated :)